subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/initialization/InstantExecutionProjectAccessListener.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.instantexecution.initialization
+
+import org.gradle.api.InvalidUserCodeException
+import org.gradle.api.Task
+import org.gradle.api.internal.GeneratedSubclasses
+import org.gradle.api.internal.project.ProjectInternal
+import org.gradle.initialization.ProjectAccessListener
+import org.gradle.instantexecution.InstantExecutionReport
+import org.gradle.instantexecution.serialization.PropertyProblem
+import org.gradle.instantexecution.serialization.PropertyTrace
+import org.gradle.instantexecution.serialization.StructuredMessage
+
+
+class InstantExecutionProjectAccessListener internal constructor(
+
+    private
+    val startParameter: InstantExecutionStartParameter,
+
+    private
+    val report: InstantExecutionReport
+
+) : ProjectAccessListener {
+
+    override fun beforeRequestingTaskByPath(targetProject: ProjectInternal) = Unit
+
+    override fun beforeResolvingProjectDependency(dependencyProject: ProjectInternal) = Unit
+
+    override fun onProjectAccess(invocationDescription: String, invocationSource: Any) {
+        if (startParameter.isEnabled) {
```

Comment:
🤔 it bothers me a bit that every `getProject()` call during task execution will have to pay the cost of delivering the event plus `startParameter.isEnabled`. Any chance this cost could be removed completely when `!startParameter.isEnabled`?

Reply:
`startParameter.isEnabled` is a `by unsafeLazy` so it should be cheap. And calls to `Task.getProject()` should not be many. No performance test complained. I would leave it like that for now and come back to it if we can observe that as a problem.

Reply:
If it's easy enough to avoid the cost by not adding the listener when `!startParameter.isEnabled`, I would. I'll leave to your judgement.